content/helpers.go

```diff
@@ -90,38 +120,65 @@ func WriteBlob(ctx context.Context, cs Ingester, ref string, r io.Reader, desc o
 // OpenWriter opens a new writer for the given reference, retrying if the writer
 // is locked until the reference is available or returns an error.
 func OpenWriter(ctx context.Context, cs Ingester, opts ...WriterOpt) (Writer, error) {
-	var (
-		cw    Writer
-		err   error
-		retry = 16
-	)
-	for {
-		cw, err = cs.Writer(ctx, opts...)
-		if err != nil {
-			if !errdefs.IsUnavailable(err) {
-				return nil, err
-			}
+	wOpt := &WriterOpts{}
+	for _, o := range opts {
+		o(wOpt)
+	}
+
+	ref := wOpt.Ref
+	if ref == "" {
+		return nil, errors.Wrap(errdefs.ErrInvalidArgument, "ref must not be empty")
+	}
 
-			// TODO: Check status to determine if the writer is active,
-			// continue waiting while active, otherwise return lock
-			// error or abort. Requires asserting for an ingest manager
+	// TODO: Keep the ref with namespace scope.
```

Comment:
Why not just scope it to the namespace in this PR?

Reply:
containerd provides namespace in two way: one is from context, and other one is from gRPC metadata context.

For CRI-plugin case, we can get namespace from context. But for client-server case, the context doesn't contains the namespace which will be provided by gRPC interceptor. 

I am still seeking common way to get the namespace :(

Reply:
Wow. I didn't know that. It would be great if we can have the namespace in the context. Alternatively we may be able to supply the namespace in WriterOpts.